src/Components/Web.JS/src/Boot.Server.Common.ts

```diff
@@ -83,42 +101,72 @@ export async function startServer(components: RootComponentManager<ServerCompone
   logger.log(LogLevel.Information, 'Blazor server-side application started.');
 
   jsInitializer.invokeAfterStartedCallbacks(Blazor);
+  resolve();
 }
 
-export function startCircuit(): Promise<boolean> {
-  if (!started) {
+export async function startCircuit(): Promise<boolean> {
+  if (!serverStartPromise) {
     throw new Error('Cannot start the circuit until Blazor Server has started.');
   }
 
-  if (circuit.didRenderingFail()) {
+  if (circuit && !circuit.isDisposedOrDisposing()) {
+    return true;
+  }
+
+  if (circuitStarting) {
+    return await circuitStarting;
+  }
+
+  await serverStartPromise;
+
+  if (circuit && circuit.didRenderingFail()) {
     // We can't start a new circuit after a rendering failure because the renderer
     // might be in an invalid state.
-    return Promise.resolve(false);
+    return false;
   }
 
-  if (circuit.isDisposedOrDisposing()) {
+  if (circuit && circuit.isDisposedOrDisposing()) {
     // If the current circuit is no longer available, create a new one.
     appState = discoverServerPersistedState(document) || '';
     circuit = new CircuitManager(circuit.getRootComponentManager(), appState, options, logger);
   }
 
   // Start the circuit. If the circuit has already started, this will return the existing
   // circuit start promise.
-  return circuit.start();
+  circuitStarting = circuit.start();
+  clearCircuitStarting(circuitStarting);
+  return circuitStarting;
+}
+
+async function clearCircuitStarting(circuitStartingPromise: Promise<boolean>) {
+  await circuitStartingPromise;
+  if (circuitStarting === circuitStartingPromise) {
+    circuitStarting = undefined;
+  }
 }
 
 export function hasStartedServer(): boolean {
-  return started;
+  return serverStartPromise !== undefined;
 }
 
 export async function disposeCircuit() {
   await circuit?.dispose();
 }
 
 export function isCircuitAvailable(): boolean {
-  return !circuit.isDisposedOrDisposing();
+  return circuit && !circuit.isDisposedOrDisposing();
 }
 
-export function updateServerRootComponents(operations: string): Promise<void>|undefined {
-  return circuit.updateRootComponents(operations);
+export function updateServerRootComponents(operations: string): Promise<void> | undefined {
+  if (circuit && !circuit.isDisposedOrDisposing()) {
+    return circuit.updateRootComponents(operations);
+  } else {
+    scheduleWhenReady(operations);
+  }
+}
+async function scheduleWhenReady(operations: string) {
+  await serverStartPromise;
+  if (await startCircuit()) {
+    return circuit.updateRootComponents(operations);
+  }
```

Comment:
When we process the document (on update root components) we fire up the server, but we don't wait for it to complete initializing, that's a problem because we later might call updateRootComponents before the server has fully initialized.

What happened before is that we "incorrectly" created the circuit synchronously in the call to startServer and relied upon that fact, so at that point, when we called updateRootComponents we had an already "initialized" circuit (on the client side) even if the circuit was still initializing on the server, so we were able to push the components (even though, they would just get buffered on the client/server until SignalR finished processing the previous request.

The "incorrect" part comes from the fact that initializers must be able to see the options for server/webassembly and are allowed to be asynchronous, which means that the circuit might not be setup by the time `startServer` returns, hence why we need to keep the promise and queue the update.